### PR TITLE
Switch DirectX Target to use the Itanium ABI

### LINKED ISSUE #110736: [DirectX] Switch to Itanium ABI
Itanium ABI is pretty much a standard for C++ ABIs and it is already used for SPIR-V generation. It uses a well-defined name mangling scheme with vendor qualifiers extension that we can use to encode HLSL resource type attributes. 

Some of the potential issues that we might encounter after the switch are:
- changed order of parameter initializations
- bitfields layout changes
The consensus is that these issues are most likely not going to be blocking and can deal with them as they arise, for example by defining the order of values in bitfields in the language spec.

DXC is using MSVC ABI. Maintaining linking compatibility between DXC and Clang libraries is not required.

### PR DESCRIPTION
**Pull request by pow2clk**
To consolidate behavior of function mangling and limit the number of places that ABI changes will need to be made, this switches the DirectX target used for HLSL to use the Itanium ABI from the Microsoft ABI. The Itanium ABI has greater flexibility in decisions regarding mangling of new types of which we have more than a few yet to add.

One effect of this will be that linking library shaders compiled with DXC will not be possible with shaders compiled with clang. That isn't considered a terribly interesting use case and one that would likely have been onerous to maintain anyway.

This involved adding a function to call all global destructors as the Microsoft ABI had done.

This requires a few changes to tests. Most notably the mangling style has changed which accounts for most of the changes. In making those changes, I took the opportunity to harmonize some very similar tests for greater consistency. I also shaved off some unneeded run flags that had probably been copied over from one test to another.

Other changes effected by using the new ABI include using different types when manipulating smaller bitfields, eliminating an unnecessary alloca in one instance in this-assignment.hlsl, changing the way static local initialization is guarded, and changing the order of inout parameters getting copied in and out. That last is a subtle change in functionality, but one where there was sufficient inconsistency in the past that standardizing is important, but the particular direction of the standardization is less important for the sake of existing shaders.

fixes #110736

### CODE REVIEW

File: clang/test/CodeGenHLSL/BasicFeatures/OutputArguments.hlsl [hunk 1]
```diff
@@ -260,10 +260,10 @@ void order_matters(inout int X, inout int Y) {
 // CHECK: store i32 [[VVal]], ptr [[Tmp0]]
 // CHECK: [[VVal:%.*]] = load i32, ptr [[V]]
 // CHECK: store i32 [[VVal]], ptr [[Tmp1]]
-// CHECK: call void {{.*}}order_matters{{.*}}(ptr noalias noundef nonnull align 4 dereferenceable(4) [[Tmp1]], ptr noalias noundef nonnull align 4 dereferenceable(4) [[Tmp0]])
-// CHECK: [[Arg1Val:%.*]] = load i32, ptr [[Tmp1]]
+// CHECK: call void {{.*}}order_matters{{.*}}(ptr noalias noundef nonnull align 4 dereferenceable(4) [[Tmp0]], ptr noalias noundef nonnull align 4 dereferenceable(4) [[Tmp1]])
+// CHECK: [[Arg1Val:%.*]] = load i32, ptr [[Tmp0]]
 // CHECK: store i32 [[Arg1Val]], ptr [[V]]
-// CHECK: [[Arg2Val:%.*]] = load i32, ptr [[Tmp0]]
+// CHECK: [[Arg2Val:%.*]] = load i32, ptr [[Tmp1]]
```
Comment (pow2clk) [inline clang/test/CodeGenHLSL/BasicFeatures/OutputArguments.hlsl:266]: This is where the order of copy in copy out is reflected

File: clang/test/CodeGenHLSL/BasicFeatures/OutputArguments.hlsl [hunk 2]
```diff
@@ -289,17 +289,19 @@ void setFour(inout int I) {
...
-// CHECK: [[BFValue:%.*]] = and i32 [[RetVal]], 255
-// CHECK: [[ZerodField:%.*]] = and i32 [[BFLoad]], -256
-// CHECK: [[BFSet:%.*]] = or i32 [[ZerodField]], [[BFValue]]
-// CHECK: store i32 [[BFSet]], ptr [[B]]
+// CHECK: [[TruncVal:%.*]] = trunc i32 [[RetVal]] to i16
+// CHECK: [[BFLoad:%.*]] = load i16, ptr [[B]]
+// CHECK: [[BFValue:%.*]] = and i16 [[TruncVal]], 255
+// CHECK: [[ZerodField:%.*]] = and i16 [[BFLoad]], -256
+// CHECK: [[BFSet:%.*]] = or i16 [[ZerodField]], [[BFValue]]
+// CHECK: store i16 [[BFSet]], ptr [[B]]
```
Comment (pow2clk) [inline clang/test/CodeGenHLSL/BasicFeatures/OutputArguments.hlsl:304]: This is an instance of using smaller types where possible. Doing so introduced some additional commands.

File: clang/test/CodeGenHLSL/builtins/abs.hlsl [hunk 3]
```diff
@@ -1,93 +1,96 @@
...
+// RUN:  FileCheck %s --check-prefixes=CHECK,NATIVE_HALF
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.3-library %s \
+// RUN:  -emit-llvm -disable-llvm-passes -o - | \
+// RUN:  FileCheck %s --check-prefixes=CHECK,NO_HALF
 
 using hlsl::abs;
 
 #ifdef __HLSL_ENABLE_16_BIT
-// NATIVE_HALF: define noundef i16 @
+// NATIVE_HALF-LABEL: define noundef i16 @_Z16test_abs_int16_t
```
Comment (pow2clk) [inline clang/test/CodeGenHLSL/builtins/abs.hlsl:11]: There was a lot of inconsistency in these sorts of checks. Some included more and some less. I tried to include enough to unambiguously identify the function in question.

File: clang/test/CodeGenHLSL/this-assignment.hlsl [hunk 4]
```diff
@@ -56,14 +55,12 @@ void main() {
...
-// CHECK-NEXT: store ptr [[AggRes:%.*]], ptr [[ResPtr]]
 // CHECK-NEXT: store ptr {{.*}}, ptr [[ThisPtrAddr]]
 // CHECK-NEXT: [[ThisPtr:%.*]] = load ptr, ptr [[ThisPtrAddr]]
 // CHECK-NEXT: call void @llvm.memcpy.p0.p0.i32(ptr align 4 [[ThisPtr]], ptr align 4 [[Obj:%.*]], i32 8, i1 false)
 // CHECK-NEXT: [[FirstAddr:%.*]] = getelementptr inbounds nuw %struct.Pair, ptr [[ThisPtr]], i32 0, i32 0
 // CHECK-NEXT: [[First:%.*]] = load i32, ptr [[FirstAddr]]
 // CHECK-NEXT: [[FirstPlusTwo:%.*]] = add nsw i32 [[First]], 2
 // CHECK-NEXT: store i32 [[FirstPlusTwo]], ptr [[FirstAddr]]
-// CHECK-NEXT: call void @llvm.memcpy.p0.p0.i32(ptr align 4 [[AggRes]], ptr align 4 [[Obj]], i32 8, i1 false)
+// CHECK-NEXT: call void @llvm.memcpy.p0.p0.i32(ptr align 4 {{.*}}, ptr align 4 [[Obj]], i32 8, i1 false)
```
Comment (pow2clk) [inline clang/test/CodeGenHLSL/this-assignment.hlsl:66]: I can't explain exactly why the `ResPtr` alloca was removed, but it was clearly unnecessary. The wildcard here represents the same location as was previously captured in `AggRes` and copied into `ResPtr`, but then `ResPtr` was never used.

File: clang/test/CodeGenHLSL/GlobalDestructors.hlsl [hunk 5]
```diff
@@ -1,7 +1,7 @@
-// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.0-compute -std=hlsl202x -emit-llvm -disable-llvm-passes %s -o - | FileCheck %s --check-prefixes=CS,NOINLINE,CHECK
-// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.3-library -std=hlsl202x -emit-llvm -disable-llvm-passes %s -o - | FileCheck %s --check-prefixes=LIB,NOINLINE,CHECK
-// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.0-compute -std=hlsl202x -emit-llvm -O0 %s -o - | FileCheck %s --check-prefixes=INLINE,CHECK
-// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.3-library -std=hlsl202x -emit-llvm -O0 %s -o - | FileCheck %s --check-prefixes=INLINE,CHECK
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.0-compute -emit-llvm -disable-llvm-passes %s -o - | FileCheck %s --check-prefixes=CS,NOINLINE,CHECK
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.3-library -emit-llvm -disable-llvm-passes %s -o - | FileCheck %s --check-prefixes=LIB,NOINLINE,CHECK
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.0-compute -emit-llvm -O0 %s -o - | FileCheck %s --check-prefixes=INLINE,CHECK
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.3-library -emit-llvm -O0 %s -o - | FileCheck %s --check-prefixes=INLINE,CHECK
```
Comment (pow2clk) [inline clang/test/CodeGenHLSL/GlobalDestructors.hlsl:4]: hlsl202x is the default now, so this flag is redundant.

File: clang/test/CodeGenHLSL/builtins/abs.hlsl [hunk 6]
```diff
@@ -1,93 +1,96 @@
...
-// RUN:   --check-prefixes=CHECK,NATIVE_HALF
-// RUN: %clang_cc1 -std=hlsl2021 -finclude-default-header -x hlsl -triple \
-// RUN:   dxil-pc-shadermodel6.3-library %s -emit-llvm -disable-llvm-passes \
-// RUN:   -o - | FileCheck %s --check-prefixes=CHECK,NO_HALF
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.3-library %s \
+// RUN:  -fnative-half-type -emit-llvm -disable-llvm-passes -o - | \
+// RUN:  FileCheck %s --check-prefixes=CHECK,NATIVE_HALF
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.3-library %s \
+// RUN:  -emit-llvm -disable-llvm-passes -o - | \
+// RUN:  FileCheck %s --check-prefixes=CHECK,NO_HALF
```
Comment (pow2clk) [inline clang/test/CodeGenHLSL/builtins/abs.hlsl:6]: The file extension already has the same effect as the flag `-x hlsl`

**Review by pow2clk — COMMENTED** [review]
Just a few comments to explain the changes that aren't just mangling style changes.

File: clang/test/CodeGenHLSL/static-local-ctor.hlsl [hunk 7]
```diff
@@ -13,16 +13,16 @@ void InitBuf(RWBuffer<int> buf) {
...
-// CHECK-NEXT: br i1 [[Tmp4]]
+// CHECK-NEXT: [[Tmp2:%.*]] = load i8, ptr @_ZGVZ4mainvE5mybuf
+// CHECK-NEXT: [[Tmp3:%.*]] = icmp eq i8 [[Tmp2]], 0
+// CHECK-NEXT: br i1 [[Tmp3]]
 // CHECK-NOT: _Init_thread_header
-// CHECK: init:
-// CHECK-NEXT: = or i32 [[Tmp2]], 1
+// CHECK: init.check:
+// CHECK-NEXT: call void @_ZN4hlsl8RWBufferIiEC1Ev
+// CHECK-NEXT: store i8 1, ptr @_ZGVZ4mainvE5mybuf
```
Comment (pow2clk) [inline clang/test/CodeGenHLSL/static-local-ctor.hlsl:25]: These changes represent a different, but equivalent way of protecting the one-time initialization of a static local variable

**Review by farzonl — APPROVED** [review]

**Review by bharadwajy — APPROVED** [review]

**Pull request merged**